File: servers/go/internal/application/commands/catalog.go
package commands

import (
	"context"
	"fmt"
	"time"

	"github.com/christophercaldwell/model-architecture/go/internal/application"
	"github.com/christophercaldwell/model-architecture/go/internal/domain/book"
	"github.com/christophercaldwell/model-architecture/go/internal/domain/bookcopy"
)

type AddBookCopyInput struct {
	ISBN    string
	Barcode string
}

type CatalogCommands struct {
	factory application.UnitOfWorkFactory
}

func NewCatalogCommands(factory application.UnitOfWorkFactory) *CatalogCommands {
	return &CatalogCommands{factory: factory}
}

func (c *CatalogCommands) getBookByISBN(ctx context.Context, uow application.UnitOfWork, isbn string) (*book.Book, error) {
	b, err := uow.Books().GetByISBN(ctx, isbn)
	if err != nil {
		return nil, fmt.Errorf("load book for write: %w", err)
	}
	if b == nil {
		return nil, book.ErrNotFound
	}
	return b, nil
}

func (c *CatalogCommands) getBookCopyByBarcode(ctx context.Context, uow application.UnitOfWork, barcode string) (*bookcopy.BookCopy, error) {
	bc, err := uow.BookCopies().GetByBarcodeForUpdate(ctx, barcode)
	if err != nil {
		return nil, fmt.Errorf("load book copy for write: %w", err)
	}
	if bc == nil {
		return nil, bookcopy.ErrNotFound
	}
	return bc, nil
}

func (c *CatalogCommands) AddBook(ctx context.Context, payload book.BookCreationPayload) (*book.Book, error) {
	prepared := payload.Prepare()
	uow, err := c.factory.New(ctx)
	if err != nil {
		return nil, fmt.Errorf("build unit of work: %w", err)
	}
	defer uow.Rollback(ctx) //nolint:errcheck

	result, err := uow.Books().Create(ctx, prepared)
	if err != nil {
		return nil, fmt.Errorf("add book: %w", err)
	}
	if err := uow.Commit(ctx); err != nil {
		return nil, fmt.Errorf("commit transaction: %w", err)
	}
	return result, nil
}

func (c *CatalogCommands) AddBookCopy(ctx context.Context, input AddBookCopyInput) (*bookcopy.BookCopy, error) {
	uow, err := c.factory.New(ctx)
	if err != nil {
		return nil, fmt.Errorf("build unit of work: %w", err)
	}
	defer uow.Rollback(ctx) //nolint:errcheck

	b, err := c.getBookByISBN(ctx, uow, input.ISBN)
	if err != nil {
		return nil, err
	}
	prepared := bookcopy.BookCopyCreationPayload{
		Barcode: input.Barcode,
		BookID:  b.ID,
	}.Prepare()
	result, err := uow.BookCopies().Create(ctx, prepared)
	if err != nil {
		return nil, fmt.Errorf("add book copy: %w", err)
	}
	if err := uow.Commit(ctx); err != nil {
		return nil, fmt.Errorf("commit transaction: %w", err)
	}
	return result, nil
}

func (c *CatalogCommands) MarkBookCopyLost(ctx context.Context, barcode string) (*bookcopy.BookCopy, error) {
	uow, err := c.factory.New(ctx)
	if err != nil {
		return nil, fmt.Errorf("build unit of work: %w", err)
	}
	defer uow.Rollback(ctx) //nolint:errcheck

	bc, err := c.getBookCopyByBarcode(ctx, uow, barcode)
	if err != nil {
		return nil, err
	}
	newStatus, err := bc.MarkLost()
	if err != nil {
		return nil, err
	}
	if err := uow.BookCopies().UpdateStatus(ctx, bc.ID, newStatus); err != nil {
		return nil, fmt.Errorf("mark book copy lost: %w", err)
	}
	if err := uow.Commit(ctx); err != nil {
		return nil, fmt.Errorf("commit transaction: %w", err)
	}
	updated := *bc
	updated.Status = newStatus
	updated.DtModified = time.Now()
	return &updated, nil
}

func (c *CatalogCommands) MarkBookCopyFound(ctx context.Context, barcode string) (*bookcopy.BookCopy, error) {
	uow, err := c.factory.New(ctx)
	if err != nil {
		return nil, fmt.Errorf("build unit of work: %w", err)
	}
	defer uow.Rollback(ctx) //nolint:errcheck

	bc, err := c.getBookCopyByBarcode(ctx, uow, barcode)
	if err != nil {
		return nil, err
	}
	newStatus, err := bc.MarkFound()
	if err != nil {
		return nil, err
	}
	if err := uow.BookCopies().UpdateStatus(ctx, bc.ID, newStatus); err != nil {
		return nil, fmt.Errorf("mark book copy found: %w", err)
	}
	if err := uow.Commit(ctx); err != nil {
		return nil, fmt.Errorf("commit transaction: %w", err)
	}
	updated := *bc
	updated.Status = newStatus
	updated.DtModified = time.Now()
	return &updated, nil
}

func (c *CatalogCommands) SendBookCopyToMaintenance(ctx context.Context, barcode string) (*bookcopy.BookCopy, error) {
	uow, err := c.factory.New(ctx)
	if err != nil {
		return nil, fmt.Errorf("build unit of work: %w", err)
	}
	defer uow.Rollback(ctx) //nolint:errcheck

	bc, err := c.getBookCopyByBarcode(ctx, uow, barcode)
	if err != nil {
		return nil, err
	}
	newStatus, err := bc.SendToMaintenance()
	if err != nil {
		return nil, err
	}
	if err := uow.BookCopies().UpdateStatus(ctx, bc.ID, newStatus); err != nil {
		return nil, fmt.Errorf("send book copy to maintenance: %w", err)
	}
	if err := uow.Commit(ctx); err != nil {
		return nil, fmt.Errorf("commit transaction: %w", err)
	}
	updated := *bc
	updated.Status = newStatus
	updated.DtModified = time.Now()
	return &updated, nil
}

func (c *CatalogCommands) CompleteBookCopyMaintenance(ctx context.Context, barcode string) (*bookcopy.BookCopy, error) {
	uow, err := c.factory.New(ctx)
	if err != nil {
		return nil, fmt.Errorf("build unit of work: %w", err)
	}
	defer uow.Rollback(ctx) //nolint:errcheck

	bc, err := c.getBookCopyByBarcode(ctx, uow, barcode)
	if err != nil {
		return nil, err
	}
	newStatus, err := bc.CompleteMaintenance()
	if err != nil {
		return nil, err
	}
	if err := uow.BookCopies().UpdateStatus(ctx, bc.ID, newStatus); err != nil {
		return nil, fmt.Errorf("complete book copy maintenance: %w", err)
	}
	if err := uow.Commit(ctx); err != nil {
		return nil, fmt.Errorf("commit transaction: %w", err)
	}
	updated := *bc
	updated.Status = newStatus
	updated.DtModified = time.Now()
	return &updated, nil
}

File: server/crates/application/src/commands/catalog.rs
use chrono::Utc;
use domain::{
    book::{Book, BookCreationPayload, BookError},
    book_copy::{BookCopy, BookCopyCreationPayload, BookCopyError},
    uow::{WriteUnitOfWork, WriteUnitOfWorkFactory},
};
use std::sync::Arc;

#[derive(Clone)]
pub struct CatalogCommands {
    uow_factory: Arc<dyn WriteUnitOfWorkFactory>,
}

impl CatalogCommands {
    #[must_use]
    pub fn new(uow_factory: Arc<dyn WriteUnitOfWorkFactory>) -> Self {
        Self { uow_factory }
    }

    async fn get_book_by_isbn(
        &self,
        uow: &mut WriteUnitOfWork,
        isbn: &str,
    ) -> Result<Book, super::CommandError> {
        uow.book()
            .get_by_isbn(isbn)
            .await?
            .ok_or(BookError::NotFound.into())
    }

    async fn get_book_copy_by_barcode(
        &self,
        uow: &mut WriteUnitOfWork,
        barcode: &str,
    ) -> Result<BookCopy, super::CommandError> {
        uow.book_copy()
            .get_by_barcode_for_update(barcode)
            .await?
            .ok_or(BookCopyError::NotFound.into())
    }

    pub async fn add_book(
        &self,
        payload: BookCreationPayload,
    ) -> Result<Book, super::CommandError> {
        let prepared = payload.prepare();
        let mut uow = self.uow_factory.build().await?;
        let result = uow.book().create(&prepared).await?;
        uow.commit().await?;
        Ok(result)
    }

    pub async fn add_book_copy(
        &self,
        input: super::AddBookCopyInput,
    ) -> Result<BookCopy, super::CommandError> {
        let mut uow = self.uow_factory.build().await?;
        let book = self.get_book_by_isbn(&mut uow, &input.isbn).await?;
        let prepared = BookCopyCreationPayload {
            barcode: input.barcode,
            book_id: book.id,
        }
        .prepare();
        let result = uow.book_copy().create(&prepared).await?;
        uow.commit().await?;
        Ok(result)
    }

    pub async fn mark_book_copy_lost(
        &self,
        barcode: String,
    ) -> Result<BookCopy, super::CommandError> {
        let mut uow = self.uow_factory.build().await?;
        let book_copy = self.get_book_copy_by_barcode(&mut uow, &barcode).await?;
        let lost_status = book_copy.mark_lost()?;
        uow.book_copy()
            .update_status(book_copy.id, lost_status.clone())
            .await?;
        uow.commit().await?;
        let updated_copy = BookCopy {
            status: lost_status,
            dt_modified: Utc::now(),
            ..book_copy
        };
        Ok(updated_copy)
    }

    pub async fn mark_book_copy_found(
        &self,
        barcode: String,
    ) -> Result<BookCopy, super::CommandError> {
        let mut uow = self.uow_factory.build().await?;
        let book_copy = self.get_book_copy_by_barcode(&mut uow, &barcode).await?;
        let found_status = book_copy.mark_found()?;
        uow.book_copy()
            .update_status(book_copy.id, found_status.clone())
            .await?;
        uow.commit().await?;
        let updated_copy = BookCopy {
            status: found_status,
            dt_modified: Utc::now(),
            ..book_copy
        };
        Ok(updated_copy)
    }

    pub async fn send_book_copy_to_maintenance(
        &self,
        barcode: String,
    ) -> Result<BookCopy, super::CommandError> {
        let mut uow = self.uow_factory.build().await?;
        let book_copy = self.get_book_copy_by_barcode(&mut uow, &barcode).await?;
        let maintenance_status = book_copy.send_to_maintenance()?;
        uow.book_copy()
            .update_status(book_copy.id, maintenance_status.clone())
            .await?;
        uow.commit().await?;
        let updated_copy = BookCopy {
            status: maintenance_status,
            dt_modified: Utc::now(),
            ..book_copy
        };
        Ok(updated_copy)
    }

    pub async fn complete_book_copy_maintenance(
        &self,
        barcode: String,
    ) -> Result<BookCopy, super::CommandError> {
        let mut uow = self.uow_factory.build().await?;
        let book_copy = self.get_book_copy_by_barcode(&mut uow, &barcode).await?;
        let active_status = book_copy.complete_maintenance()?;
        uow.book_copy()
            .update_status(book_copy.id, active_status.clone())
            .await?;
        uow.commit().await?;
        let updated_copy = BookCopy {
            status: active_status,
            dt_modified: Utc::now(),
            ..book_copy
        };
        Ok(updated_copy)
    }
}
